lighthouse-core/audits/seo/robots-txt.js

```diff
@@ -0,0 +1,195 @@
+ */
+'use strict';
+
+const Audit = require('../audit');
+const URL = require('../../lib/url-shim');
+
+const HTTP_CLIENT_ERROR_CODE_LOW = 400;
+const HTTP_SERVER_ERROR_CODE_LOW = 500;
+
+const DIRECTIVE_SITEMAP = 'sitemap';
+const DIRECTIVE_USER_AGENT = 'user-agent';
+const DIRECTIVE_ALLOW = 'allow';
+const DIRECTIVE_DISALLOW = 'disallow';
+const DIRECTIVES_GROUP_MEMBERS = new Set([DIRECTIVE_ALLOW, DIRECTIVE_DISALLOW]);
+const DIRECTIVE_SAFELIST = new Set([
+  DIRECTIVE_USER_AGENT, DIRECTIVE_DISALLOW, // standard
+  DIRECTIVE_ALLOW, DIRECTIVE_SITEMAP, // universally supported
+  'crawl-delay', // yahoo, bing, yandex
+  'clean-param', 'host', // yandex
+  'request-rate', 'visit-time', 'noindex', // not officially supported, but used in the wild
+]);
+const SITEMAP_VALID_PROTOCOLS = new Set(['https:', 'http:', 'ftp:']);
+
+/**
+ * @param {!string} line single line from a robots.txt file
+ * @returns {!{directive: string, value: string}}
+ */
+function parseLine(line) {
+  const hashIndex = line.indexOf('#');
+
+  if (hashIndex !== -1) {
+    line = line.substr(0, hashIndex);
+  }
+
+  line = line.trim();
+
+  if (line.length === 0) {
+    return {};
+  }
+
+  const colonIndex = line.indexOf(':');
+
+  if (colonIndex === -1) {
+    throw new Error('Syntax not understood');
+  }
+
+  const directiveName = line.slice(0, colonIndex).trim().toLowerCase();
+  const directiveValue = line.slice(colonIndex + 1).trim();
+
+  if (!DIRECTIVE_SAFELIST.has(directiveName)) {
+    throw new Error('Unknown directive');
+  }
+
+  if (directiveName === DIRECTIVE_SITEMAP) {
+    let sitemapUrl;
+
+    try {
+      sitemapUrl = new URL(directiveValue);
+    } catch (e) {
+      throw new Error('Invalid sitemap URL');
+    }
+
+    if (!SITEMAP_VALID_PROTOCOLS.has(sitemapUrl.protocol)) {
+      throw new Error('Invalid sitemap URL protocol');
+    }
+  }
+
+  if (directiveName === DIRECTIVE_USER_AGENT && !directiveValue) {
+    throw new Error('No user-agent specified');
+  }
+
+  if (directiveName === DIRECTIVE_ALLOW || directiveName === DIRECTIVE_DISALLOW) {
+    if (directiveValue !== '' && directiveValue[0] !== '/' && directiveValue[0] !== '*') {
+      throw new Error('Pattern should either be empty, start with "/" or "*"');
+    }
+
+    const dolarIndex = directiveValue.indexOf('$');
+
+    if (dolarIndex !== -1 && dolarIndex !== directiveValue.length - 1) {
+      throw new Error('"$" should only be used at the end of the pattern');
+    }
+  }
+
+  return {
+    directive: directiveName,
+    value: directiveValue,
+  };
+}
+
+function validateRobots(content) {
+  let inGroup = false;
+
+  return content
+    .split(/\r\n|\r|\n/)
+    .map((line, index) => {
+      let parsedLine;
+
+      try {
+        parsedLine = parseLine(line);
+      } catch (e) {
+        return {
+          index: index + 1,
+          line: line,
+          error: e.message,
+        };
+      }
+
+      if (parsedLine.directive === DIRECTIVE_USER_AGENT) {
+        inGroup = true;
```

Comment:
once you enter a group you never leave it? man robots files are weird 😆 

a comment or two explaining how this works would be nice for us less seo-expert maintainers :)

Reply:
I added a short comment with explanation and doc link in this code 👍 

BTW I wrote down list of rules and resources that I've used [here](https://github.com/GoogleChrome/lighthouse/issues/4356#issuecomment-375489925)